RavenDB-13556 Improve client interaction with the cluster

karmeli87 opened the pull request:
This PR touch many files, so I've split it into several logical commits.

The most interesting one, is probably the "core changes".

karmeli87 committed:
RavenDB-13556 Improve client interaction with the cluster

implement raft unique request id client side commands

karmeli87 committed:
RavenDB-13556 Improve client interaction with the cluster

implement unique request id for all raft command (server-side)

karmeli87 committed:
RavenDB-13556 Improve client interaction with the cluster

modify handlers and functions to pass the unique request id (server-side)

karmeli87 committed:
RavenDB-13556 Improve client interaction with the cluster

modify test to pass the unique request id

karmeli87 committed:
RavenDB-13556 Improve client interaction with the cluster

core changes

karmeli87 committed:
RavenDB-13556 Improve client interaction with the cluster

fix missing commands and failing tests

ayende commented on src/Raven.Client/Http/RequestExecutor.cs:
```diff
@@ -1025,11 +1025,21 @@ private HttpRequestMessage CreateRequest<TResult>(JsonOperationContext ctx, Serv
-                request.Headers.Add(Constants.Headers.ClientVersion, ClientVersion);
+            if (command is IRaftCommand raftCommand)
+            {
+                Debug.Assert(raftCommand.RaftUniqueRequestId != null, $"Forget to create an id for {command.GetType()}?");
```
Also check for empty?

karmeli87 replied:
I keep the empty for "don't care"

ayende commented on src/Raven.Client/Http/RequestExecutor.cs:
```diff
@@ -1025,11 +1025,21 @@ private HttpRequestMessage CreateRequest<TResult>(JsonOperationContext ctx, Serv
+                Debug.Assert(raftCommand.RaftUniqueRequestId != null, $"Forget to create an id for {command.GetType()}?");
+
+                var raftRequestString = "raft-request-id=" + raftCommand.RaftUniqueRequestId;
+                builder.Query = builder.Query?.Length > 1 ? $"{builder.Query.Substring(1)}&{raftRequestString}" : raftRequestString;
```
Why are you skipping the first char in the existing `Query` ?

karmeli87 replied:
followed this : https://docs.microsoft.com/en-us/dotnet/api/system.uribuilder.query?view=netframework-4.8

ayende commented on src/Raven.Server/Commercial/LicenseManager.cs:
```diff
@@ -429,7 +429,7 @@ private void AssertCanAssignCores(int assignedCores)
         }
 
-        public async Task Activate(License license, bool skipLeaseLicense, bool ensureNotPassive = true, bool forceActivate = false)
+        public async Task Activate(License license, bool skipLeaseLicense, string guid, bool ensureNotPassive = true, bool forceActivate = false)
```
`guid` of what? It looks like this related to the license id. Better to name it properly

ayende commented on src/Raven.Server/Commercial/LicenseManager.cs:
```diff
@@ -703,7 +704,7 @@ private async Task ExecuteTasks()
         }
 
-        public async Task LeaseLicense(bool forceUpdate = false)
+        public async Task LeaseLicense(string guid, bool forceUpdate = false)
```
```suggestion
        public async Task LeaseLicense(string raftReqId, bool forceUpdate = false)
```

Or something like that

ayende commented on src/Raven.Server/Documents/DatabaseRequestHandler.cs:
```diff
@@ -44,9 +44,11 @@ public override void Init(RequestHandlerContext context)
+
+        protected async Task DatabaseConfigurations(SetupFunc setupConfigurationFunc,
            string debug,
+           string guid,
```
Meaningful name

ayende commented on src/Raven.Server/Documents/Handlers/IndexHandler.cs:
```diff
@@ -638,14 +639,15 @@ public async Task SetLockMode()
                 foreach (var name in parameters.IndexNames)
                 {
-                    await Database.IndexStore.SetPriority(name, parameters.Priority);
+                    await Database.IndexStore.SetPriority(name, parameters.Priority, $"{guid}/{name}");
```
Are they guaranteed to be unique? Also elsewhere where you do this

karmeli87 replied:
guid is unique.. as long as the name of the index is also unique (or any other suffix for the particular operation), it should be fine.

ayende replied:
I don't think we _require_ that the name will be unique, though.
Better to add an index number there too

karmeli87 replied:
Would it make sense or even possible to pass 2 indexes with the same name in the same operation?

ayende replied:
Sense, no? 
Possible...

ayende commented on src/Raven.Server/Documents/PeriodicBackup/Restore/RestoreBackupTask.cs:
```diff
@@ -180,7 +180,8 @@ public async Task<IOperationResult> Execute(Action<IOperationProgress> onProgres
                         databaseRecord.DatabaseState = DatabaseStateStatus.RestoreInProgress;
 
-                        var (index, _) = await _serverStore.WriteDatabaseRecordAsync(databaseName, databaseRecord, null, restoreSettings.DatabaseValues, isRestore: true);
+                        // TODO: does string.Empty make sense here?
```
Probably not, get the req id from the top?

ayende commented on src/Raven.Server/Rachis/Leader.cs:
```diff
@@ -710,19 +710,41 @@ private void EmptyQueue()
+                            {
+                                if (exception != null)
+                                {
+                                    cmd.Tcs.TrySetException(new Exception(exception));
```
I don't like this, however. You are throwing `Exception`, and we probably should be throwing the actual exception type that was raised.

karmeli87 replied:
Agree, but I keep the exception as a string and not sure how to convert it back to the original one.

ayende replied:
Keep the exception time and create it dynamically?

ayende commented on src/Raven.Server/Rachis/RachisConsensus.cs:
```diff
@@ -282,9 +284,29 @@ private set
         internal static readonly Slice EntriesSlice;
         internal static readonly TableSchema LogsTable;
+        internal static readonly TableSchema LogHistoryTable;
+        private const int LogHistoryMaxEntries = 2048;
```
This should be configurable

ayende commented on src/Raven.Server/Rachis/RachisConsensus.cs:
```diff
@@ -282,9 +284,29 @@ private set
+        internal static readonly TableSchema LogHistoryTable;
+        private const int LogHistoryMaxEntries = 2048;
+        internal enum LogHistoryColumn
+        {
```
For debug purposes, also include the time.
We can also expose this as an admin debug endpoint, which can be really useful.

ayende commented on src/Raven.Server/Web/RequestHandler.cs:
```diff
@@ -250,6 +250,27 @@ protected bool IsWebsocketRequest()
+
+            if (RequestRouter.TryGetClientVersion(HttpContext, out var clientVersion))
+            {
+                if (clientVersion.Build <= 42_008) // TODO: what should be the behaviour with the nightly version 
```
The nightly also have a version indication, no?

karmeli87 replied:
IIRC nightlys are simply 42

ppekrol replied:
Should be `42_009` ?

ayende commented on src/Raven.Server/Web/RequestHandler.cs:
```diff
@@ -250,6 +250,27 @@ protected bool IsWebsocketRequest()
+            }
+
+#if DEBUG
+            return guid;
```
I don't understand why you have this?

ayende replied:
If the version match, and no request id provided, should throw

karmeli87 replied:
I want to avoid throwing any exception in `release`. 
In order not to break any client..

ayende replied:
Right now you aren't throwing anything anyway, though?

karmeli87 replied:
If the guid is null, I do throw before appending to the log (both release and debug)
I want to throw only in debug builds, in release even if we have a bug we should put `string.Empty`.

karmeli87 committed:
RavenDB-13556 Improve client interaction with the cluster

address PR comments

ppekrol commented on src/Raven.Server/Rachis/RachisConsensus.cs:
```diff
@@ -313,6 +342,225 @@ static RachisConsensus()
+            var djv = new DynamicJsonValue();
+
+            var ticks = Bits.SwapBytes(*(long*)entryHolder.Reader.Read((int)(LogHistoryColumn.Ticks), out _));
+            djv["Date"] = new DateTime(ticks);
```
nameof

ppekrol commented on src/Raven.Server/Rachis/RachisConsensus.cs:
```diff
@@ -313,6 +342,225 @@ static RachisConsensus()
+            djv["Date"] = new DateTime(ticks);
+
+            var guidPtr = entryHolder.Reader.Read((int)(LogHistoryColumn.Guid), out var size);
+            djv[LogHistoryColumn.Guid.ToString()] = Encoding.UTF8.GetString(guidPtr, size);
```
nameof

ppekrol commented on src/Raven.Server/Rachis/RachisConsensus.cs:
```diff
@@ -313,6 +342,225 @@ static RachisConsensus()
+            djv[LogHistoryColumn.Guid.ToString()] = Encoding.UTF8.GetString(guidPtr, size);
+
+            var index = Bits.SwapBytes(*(long*)entryHolder.Reader.Read((int)(LogHistoryColumn.Index), out _));
+            djv[LogHistoryColumn.Index.ToString()] = index;
```
nameof

ppekrol commented on src/Raven.Server/Rachis/RachisConsensus.cs:
```diff
@@ -313,6 +342,225 @@ static RachisConsensus()
+            var index = Bits.SwapBytes(*(long*)entryHolder.Reader.Read((int)(LogHistoryColumn.Index), out _));
+            djv[LogHistoryColumn.Index.ToString()] = index;
+
+            djv[LogHistoryColumn.Term.ToString()] = *(long*)entryHolder.Reader.Read((int)(LogHistoryColumn.Term), out _);
```
nameof

ppekrol commented on src/Raven.Server/Rachis/RachisConsensus.cs:
```diff
@@ -313,6 +342,225 @@ static RachisConsensus()
+            djv[LogHistoryColumn.Term.ToString()] = *(long*)entryHolder.Reader.Read((int)(LogHistoryColumn.Term), out _);
+
+            var typeString = entryHolder.Reader.Read((int)(LogHistoryColumn.Type), out size);
+            djv[LogHistoryColumn.Type.ToString()] = Encoding.UTF8.GetString(typeString, size);
```
nameof

ppekrol commented on src/Raven.Server/Rachis/RachisConsensus.cs:
```diff
@@ -313,6 +342,225 @@ static RachisConsensus()
+            var typeString = entryHolder.Reader.Read((int)(LogHistoryColumn.Type), out size);
+            djv[LogHistoryColumn.Type.ToString()] = Encoding.UTF8.GetString(typeString, size);
+
+            djv[LogHistoryColumn.State.ToString()] = (*(HistoryStatus*)entryHolder.Reader.Read((int)(LogHistoryColumn.State), out _)).ToString();
```
nameof

ppekrol commented on src/Raven.Server/Rachis/RachisConsensus.cs:
```diff
@@ -313,6 +342,225 @@ static RachisConsensus()
+            if (size > 0)
+            {
+                var blittableResult = new BlittableJsonReaderObject(resultPtr, size, context);
+                djv[LogHistoryColumn.Result.ToString()] = blittableResult.ToString();
```
nameof

ppekrol commented on src/Raven.Server/Rachis/RachisConsensus.cs:
```diff
@@ -313,6 +342,225 @@ static RachisConsensus()
+            }
+            else
+            {
+                djv[LogHistoryColumn.Result.ToString()] = null;
```
nameof

ppekrol commented on src/Raven.Server/Rachis/RachisConsensus.cs:
```diff
@@ -313,6 +342,225 @@ static RachisConsensus()
+            }
+
+            var exTypePtr = entryHolder.Reader.Read((int)(LogHistoryColumn.ExceptionType), out size);
+            djv[LogHistoryColumn.ExceptionType.ToString()] = size > 0 ? Encoding.UTF8.GetString(exTypePtr, size) : null;
```
nameof

ppekrol commented on src/Raven.Server/Rachis/RachisConsensus.cs:
```diff
@@ -313,6 +342,225 @@ static RachisConsensus()
+            djv[LogHistoryColumn.ExceptionType.ToString()] = size > 0 ? Encoding.UTF8.GetString(exTypePtr, size) : null;
+
+            var exMsg = entryHolder.Reader.Read((int)(LogHistoryColumn.ExceptionMessage), out size);
+            djv[LogHistoryColumn.ExceptionMessage.ToString()] = size > 0 ? Encoding.UTF8.GetString(exMsg, size) : null;
```
nameof

ppekrol commented on src/Raven.Server/ServerWide/Commands/DeleteDatabaseCommand.cs:
```diff
@@ -14,12 +14,11 @@ public class DeleteDatabaseCommand : UpdateDatabaseCommand
         public bool UpdateReplicationFactor = true;
 
-        public DeleteDatabaseCommand() : base(null)
+        public DeleteDatabaseCommand()
```
ErrorOnDatabaseDoesNotExists will be `false` now?

karmeli87 replied:
This `ctor` is only for de-serialization. No code path uses it. 
What we do missing is to include the `ErrorOnDatabaseDoesNotExists ` in the `ToJson` of the `UpdateDatabaseCommand`

ppekrol replied:
Depends for what purpose it is used. You know better.

ppekrol commented on src/Raven.Client/Documents/Commands/Batches/BatchCommand.cs:
```diff
@@ -159,5 +159,7 @@ public void Dispose()
                 command?.Dispose();
         }
+
+        public string RaftUniqueRequestId { get; } = Guid.NewGuid().ToString();
```
Since we are using that all over the place and there are some magic string here e.g. `string.Empty`. I would say that we need to create internal static class called `RaftIdentifier` with two methods:
1) `NewId` which will return this
2) `OtherNameThatExplainsWhatItDoes` which will return `string.Empty`

ppekrol replied:
This will also help centralize the identifiers.

ppekrol commented on src/Raven.Client/Documents/Commands/DropSubscriptionConnectionCommand.cs:
```diff
@@ -4,14 +4,17 @@
             _name = name;
         }
+
+        
```
what happened here? :)

karmeli87 committed:
RavenDB-13556 Improve client interaction with the cluster

address PR comments:

1. create class for generating the raft request ids
2. certificate handler will rely also on the raft request id
3. create ClusterEngineVersion for snapshot validation

karmeli87 committed:
RavenDB-13556 Improve client interaction with the cluster

use a method instead of property

ppekrol committed:
RavenDB-13556 moved RaftIdGenerator to separate file and marked as internal

karmeli87 committed:
RavenDB-13556 create a seperate class for the rachis history log

set log entries as cancel if the follower need to rewind

karmeli87 committed:
RavenDB-13556 compare exchage return value is now a class instead of a tuple

ppekrol commented:
Rebased.
